### PR TITLE
APIRule CR technical reference docs

### PR DESCRIPTION
**Pull request by videlov**
**Description**

Changes proposed in this pull request:

- APIRule CR technical reference docs

**Related issue(s)**

https://github.com/kyma-project/api-gateway/issues/314

### CODE REVIEW

File: docs/api-rule-cr.md [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+    name: httpbin
+    port: 8000
+  rules:
+    - path: /headers
+      methods: ["GET"]
+      mutators:
+        - handler: header
+          config:
+            headers:
+              # Add a new header called X-Custom-Auth with the value of the incoming Authorization header
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:235]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+              X-Custom-Auth: "%REQ(Authorization)%"
+              # Add a new header called X-Some-Data with the value "some-data"
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:237]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+            authentications:
+              - issuer: $ISSUER
+                jwksUri: $JWKS_URI
+```
+
+  </details>
+</div>
+
+#### Cookie mutator
+The cookies are specified via the `cookies` field of the cookie mutator configuration field. The keys are the names of the cookies and the values are a string. In the cookie value it is possible to use [Envoy command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators). The configured cookies are set as `Cookie`-header in the request and overwrite an existing `Cookie`-header.
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:251]:
```suggestion
The cookies are specified in the **cookies** field of the cookie mutator configuration field. The keys are the names of the cookies, and each value is a string. In the cookie value, it is possible to use [Envoy command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators). The configured cookies are set as `Cookie`-header in the request and overwrite an existing one.
```

File: docs/api-rule-cr.md [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    name: httpbin
+    port: 8000
+  rules:
+    - path: /headers
+      methods: ["GET"]
+      mutators:
+        - handler: cookie
+          config:
+            cookies:
+              # Add a new cookie called some-data with the value "data"
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:278]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+              - issuer: $ISSUER
+                jwksUri: $JWKS_URI
+```
+
+  </details>
+</div>
+
+## Additional information
+
+When you fetch an existing APIRule CR, the system adds the **status** section which describes the status of the VirtualService and the Oathkeeper Access Rule created for this CR. This table lists the fields of the **status** section.
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:293]:
```suggestion
When you fetch an existing APIRule CR, the system adds the **status** section which describes the status of the VirtualService and the Oathkeeper Access Rule created for this CR. The following table lists the fields of the **status** section.
```

File: docs/api-rule-cr.md [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+>**CAUTION:** If `service` is not defined at **spec.service** level, all defined rules must have `service` defined at **spec.rules.service** level, otherwise the validation fails.
+
+>**CAUTION:** We do not support having both Oathkeeper and Istio `jwt` access strategies at the same time. Access strategies `noop` or `allow` can **not** be used with any other on the same `spec.rules.path`.
+
+### JWT access strategy
+
+>**CAUTION:** Istio JWT is **not** production ready feature and API can change
+
+#### Enabling Istio JWT
+
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:106]:
```suggestion

To enable Istio JWT, run the following command:

```

File: docs/api-rule-cr.md [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+``` sh
+kubectl patch configmap/api-gateway-config -n kyma-system --type merge -p '{"data":{"api-gateway-config":"jwtHandler: istio"}}'
+```
+
+#### Switching back to Oathkeeper JWT
+
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:112]:
```suggestion

To enable Oathkeeper JWT, run the following command:

```

File: docs/api-rule-cr.md [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+| **spec.rules.accessStrategies.config.authentications.fromHeaders**        | List of headers from which the JWT token is taken.                         |
+| **spec.rules.accessStrategies.config.authentications.fromHeaders.name**   | Name of the header.                                                        |
+| **spec.rules.accessStrategies.config.authentications.fromHeaders.prefix** | Prefix used before the JWT token. The default is `Bearer `.                |
+| **spec.rules.accessStrategies.config.authentications.fromParams**         | List of parameters from which the JWT token is taken.                      |
+| **spec.rules.accessStrategies.config.authorizations**                     | List of authorization objects.                                             |
+| **spec.rules.accessStrategies.config.authorizations.requiredScopes**      | List of required scope values for the JWT.                                 |
+| **spec.rules.accessStrategies.config.authorizations.audiences**           | List of audiences required for the JWT.                                    |
+
+>**CAUTION:** You can define multiple **but** different JWT issuers.
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:134]:
```suggestion
>**CAUTION:** You can define multiple JWT issuers, but each of them must be unique.
```

File: docs/api-rule-cr.md [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+      methods: ["GET"]
+      mutators: []
+      accessStrategies:
+        - handler: jwt
+          config:
+            authentications:
+            - issuer: $ISSUER
+              jwksUri: $JWKS_URI
+              fromHeaders:
+              # JWT token must be taken from a HTTP header called X-JWT-Assertion and it will have a "Kyma " prefix
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:167]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+              - name: X-JWT-Assertion
+                prefix: "Kyma "
+            - issuer: $ISSUER2
+              jwksUri: $JWKS_URI2
+              fromParameters:
+              # JWT token must be taken from URL parameter called jwt_token
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:173]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+              - "jwt_token"
+            authorizations:
+            # Allow only JWTs with the claim "scp", "scope" or "scopes" with the value "test" and the audience "example.com" and "example.org"
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:176]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+            # or JWTs with the claim "scp", "scope" or "scopes" with the values "read" and "write"
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:177]:
```suggestion
```

File: docs/api-rule-cr.md [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+>**CAUTION:** You can define multiple **but** different JWT issuers.
+
+>**CAUTION:** Currently, we support only a single `fromHeader` **or** a single `fromParameter`. Specifying both of these fields for a JWT issuer is not supported.
+
+<div tabs name="api-rule" group="sample-cr">
+  <details>
+  <summary label="Example">
+  Example
+  </summary>
+
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:143]:
```suggestion

In the following example, the APIRule has two defined issuers. `ISSUER` uses a JWT token taken from the HTTP header, which is called `X-JWT-Assertion` and has a `Kyma ` prefix. `ISSUER2` uses a JWT token taken from the URL parameter, which is called `jwt-token`. The **requiredScopes** defined in the **authorizations** field allow only JWTs with the claims `scp`, `scope` or `scopes`, the value `test`, and the audiences `example.com` and `example.org` or JWTs with the claims `scp`, `scope` or `scopes` and the values `read` and `write`.

```

File: docs/api-rule-cr.md [hunk 14]
```diff
@@ -0,0 +1,311 @@
...
+#### Header mutator
+The headers are specified via the `headers` field of the header mutator configuration field. The keys are the names of the headers and the values are a string. In the header value it is possible to use [Envoy command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators), e.g. to write an incoming header to a new header. The configured headers are set to the request and overwrite all existing headers with the same name.
+
+<div tabs name="api-rule" group="sample-cr">
+  <details>
+  <summary label="Example">
+  Example
+  </summary>
+
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:215]:
```suggestion

In the following example, two different headers are specified: **X-Custom-Auth**, which uses the incoming Authorization header as a value, and **X-Some-Data** with the value `some-data`.

```

File: docs/api-rule-cr.md [hunk 15]
```diff
@@ -0,0 +1,311 @@
...
+#### Cookie mutator
+The cookies are specified via the `cookies` field of the cookie mutator configuration field. The keys are the names of the cookies and the values are a string. In the cookie value it is possible to use [Envoy command operators](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage#command-operators). The configured cookies are set as `Cookie`-header in the request and overwrite an existing `Cookie`-header.
+
+<div tabs name="api-rule" group="sample-cr">
+  <details>
+  <summary label="Example">
+  Example
+  </summary>
+
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:258]:
```suggestion

The following APIRule example has a new cookie added with the name `some-data` and the value `data`.

```

File: docs/api-rule-cr.md [hunk 16]
```diff
@@ -0,0 +1,311 @@
...
+| **spec.rules.mutators** | **NO** | Specifies the list of [Oathkeeper](https://www.ory.sh/docs/next/oathkeeper/pipeline/mutator) or Istio mutators. |
+| **spec.rules.accessStrategies** | **YES** | Specifies the list of access strategies. Supported are [Oathkeeper](https://www.ory.sh/docs/next/oathkeeper/pipeline/authn) `oauth2_introspection`, `jwt`, `noop` and `allow`. We also support `jwt` as [Istio](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt/) access strategy. |
+
+>**CAUTION:** If `service` is not defined at **spec.service** level, all defined rules must have `service` defined at **spec.rules.service** level, otherwise the validation fails.
+
+>**CAUTION:** We do not support having both Oathkeeper and Istio `jwt` access strategies at the same time. Access strategies `noop` or `allow` can **not** be used with any other on the same `spec.rules.path`.
+
+### JWT access strategy
+
+>**CAUTION:** Istio JWT is **not** production ready feature and API can change
```
Comment (strekm) [inline docs/api-rule-cr.md:103]: please move this note to `Istio JWT configuration`

Comment (videlov) [reply docs/api-rule-cr.md]: Done.

File: docs/api-rule-cr.md [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+
+>**CAUTION:** Currently, we support only a single `fromHeader` **or** a single `fromParameter`. Specifying both of these fields for a JWT issuer is not supported.
+
+<div tabs name="api-rule" group="sample-cr">
+  <details>
+  <summary label="Example">
+  Example
+  </summary>
+
+In the following example, the APIRule has two defined issuers. `ISSUER` uses a JWT token taken from the HTTP header, which is called `X-JWT-Assertion` and has a `Kyma ` prefix. `ISSUER2` uses a JWT token taken from the URL parameter, which is called `jwt-token`. The **requiredScopes** defined in the **authorizations** field allow only JWTs with the claims `scp`, `scope` or `scopes`, the value `test`, and the audiences `example.com` and `example.org` or JWTs with the claims `scp`, `scope` or `scopes` and the values `read` and `write`.
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:148]:
```suggestion
In the following example, the APIRule has two defined Issuers. `ISSUER` uses a JWT token taken from the HTTP header, which is called `X-JWT-Assertion` and has a `Kyma ` prefix. `ISSUER2` uses a JWT token taken from the URL parameter, which is called `jwt-token`. The **requiredScopes** defined in the **authorizations** field allow only JWTs with the claims `scp`, `scope` or `scopes`, the value `test`, and the audiences `example.com` and `example.org` or JWTs with the claims `scp`, `scope` or `scopes` and the values `read` and `write`.
```

File: docs/api-rule-cr.md [hunk 18]
```diff
@@ -0,0 +1,314 @@
+---
+title: API Rule
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:2]:
```suggestion
title: APIRule
```

File: docs/api-rule-cr.md [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+
+The `apirules.gateway.kyma-project.io` CustomResourceDefinition (CRD) describes the kind and the format of data the API Gateway Controller listens for. To get the up-to-date CRD in the `yaml` format, run the following command:
+
+```shell
+kubectl get crd apirules.gateway.kyma-project.io -o yaml
+```
+
+## Sample custom resource
+
+This is a sample custom resource (CR) that the API Gateway Controller listens for to expose a service. The following example has the **rules** section specified which makes the API Gateway Controller create an Oathkeeper Access Rule for the service.
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:13]:
```suggestion
This is a sample custom resource (CR) that the API Gateway Controller listens for to expose a service. The following example has the **rules** section specified which makes API Gateway Controller create an Oathkeeper Access Rule for the service.
```

File: docs/api-rule-cr.md [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+|:---|:---|
+| **status.apiRuleStatus** | Status code describing the APIRule CR. |
+| **status.virtualServiceStatus.code** | Status code describing the VirtualService. |
+| **status.virtualService.desc** | Current state of the VirtualService. |
+| **status.accessRuleStatus.code** | Status code describing the Oathkeeper Rule. |
+| **status.accessRuleStatus.desc** | Current state of the Oathkeeper Rule. |
+
+### Status codes
+
+These are the status codes used to describe the VirtualServices and Oathkeeper Access Rules:
```
Comment (nataliasitko) [inline docs/api-rule-cr.md:308]:
```suggestion
These are the status codes used to describe VirtualServices and Oathkeeper Access Rules:
```

**Review by nataliasitko — APPROVED** [review]

**Pull request merged**